### PR TITLE
Improvements to diag info

### PR DESCRIPTION
**Pull request by luiscantero**
* Reformat diag info for easier readability and consistency
* Added units to diag info values
* Corrected term "outgress" to "egress" including option em|maxegressmessagequeue
* Changed default egress queue from 20 (max 50 MB) to 8192 (max 2 GB), as it behaved better during local debugging in VS, otherwise too many messages are dropped
* Some cosmetic improvements

image

### CODE REVIEW

**Comment by karok2m** [comment]
What's the difference between `#`, `|` and `v` characters in front of the metrics ?

File: api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/EngineConfigurationApiModel.cs [hunk 1]
```diff
@@ -43,10 +43,10 @@ public class EngineConfigurationApiModel {
         public int? MaxMessageSize { get; set; }
 
         /// <summary>
-        /// Max IoT D2C outgress message buffer size
+        /// Max IoT D2C egress message queue
         /// </summary>
-        [DataMember(Name = "maxOutgressMessages", Order = 4,
+        [DataMember(Name = "maxEgressMessages", Order = 4,
```
Comment (karok2m) [inline api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/EngineConfigurationApiModel.cs:48]: Please use the same name for data member as for class member, i.e. `maxEgressMessageQueue`.

**Comment by luiscantero** [comment]
> What's the difference between `#`, `|` and `v` characters in front of the metrics ?

It's just cosmetic, the | and the v are supposed to look like an arrow pointing down and they group 3 sections, ingress, encoding and egress. The # is for data that doesn't belong to any of the categories.

Do you think it's confusing, too noisy? any suggestions to improve it?

File: api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/EngineConfigurationApiModel.cs (hunk 1 above)
Comment (luiscantero) [reply api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/EngineConfigurationApiModel.cs]: Fixed

File: modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Runtime/LegacyCliOptions.cs [hunk 2]
```diff
@@ -294,7 +294,7 @@ public class LegacyCliOptions : Dictionary<string, string>, IAgentConfigProvider
                 BatchTriggerInterval = GetValueOrDefault<TimeSpan>(LegacyCliConfigKeys.BatchTriggerInterval, TimeSpan.FromSeconds(10)),
                 MaxMessageSize = GetValueOrDefault(LegacyCliConfigKeys.MaxMessageSize, 0),
                 ScaleTestCount = GetValueOrDefault(LegacyCliConfigKeys.ScaleTestCount, 1),
-                MaxOutgressMessages = GetValueOrDefault(LegacyCliConfigKeys.MaxOutgressMessages, 200)
+                MaxEgressMessageQueue = GetValueOrDefault(LegacyCliConfigKeys.MaxEgressMessageQueue, 8192) // 8192 * 256 KB = 2 GB.
```
Comment (koepalex) [inline modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Runtime/LegacyCliOptions.cs:297]: >8192 [](start = 101, length = 4)

2GB ? this is all ram my docker container will get, if the service is running at 400MB avgerage memory, why collecting factor 5 the size as messages? It is very unlikley that they will ever send out. What about 1 GB as default?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Publisher/src/Services/PublisherJobService.cs [hunk 3]
```diff
@@ -49,15 +49,15 @@ public sealed class PublisherJobService : IPublishServices<string> {
         });
 
         /// <summary>
-        /// Read default max outgress message buffer size from environment
+        /// Read default max egress message queue size from environment
         /// </summary>
-        internal Lazy<int> DefaultMaxOutgressMessages => new Lazy<int>(() => {
+        internal Lazy<int> DefaultMaxEgressMessageQueue => new Lazy<int>(() => {
             var env = Environment.GetEnvironmentVariable("PCS_DEFAULT_PUBLISH_MAX_OUTGRESS_MESSAGES");
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Publisher/src/Services/PublisherJobService.cs:55]: >"PCS_DEFAULT_PUBLISH_MAX_OUTGRESS_MESSAGES" [](start = 57, length = 43)

If was as bugfix implemented (quick and not so clean) when we reqork we should add the environment variable name to PcsVariables constant class.
Why didn't you also changed the environment variable? (it is not yet released)

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/MonitoredItemMessageEncoder.cs [hunk 4]
```diff
@@ -88,7 +97,6 @@ public class MonitoredItemMessageEncoder : IMessageEncoder {
             var current = notifications.GetEnumerator();
             var processing = current.MoveNext();
             var messageSize = 2; // array brackets
-            maxMessageSize -= 2048; // reserve 2k for header
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/MonitoredItemMessageEncoder.cs:91]: why deleted this line? we need to reserve space for header in payload

**Comment by koepalex** [comment]
>         private IEnumerable<NetworkMessageModel> EncodeBatchAsUadp(

[Minor] maybe the right time to refactor to avoid doublicate code in encode time

---
Refers to: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/MonitoredItemMessageEncoder.cs:167 in 2751cbf. [](commit_id = 2751cbf9cc8214a00ef8ed9f242b505ccc676e46, deletion_comment = False)

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs [hunk 5]
```diff
@@ -102,19 +103,19 @@ public class DataFlowProcessingEngine : IProcessingEngine, IDisposable {
                         }
                     },
                     new ExecutionDataflowBlockOptions {
-                        CancellationToken = cancellationToken
+                        CancellationToken = cancellationToken,
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs:106]: >, [](start = 61, length = 1)

please remove the comma if there is only one parameter used

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs [hunk 6]
```diff
@@ -89,9 +97,8 @@ public class NetworkMessageEncoder : IMessageEncoder {
             var current = notifications.GetEnumerator();
             var processing = current.MoveNext();
             var messageSize = 2; // array brackets
-            maxMessageSize -= 2048; // reserve 2k for header
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs:92]: >maxMessageSize -= 2048; [](start = 12, length = 23)

why deleted this line? we need to reserve space for header in payload

**Comment by koepalex** [comment]
>         private IEnumerable<NetworkMessageModel> EncodeBatchAsUadp(

[Minor] maybe the right time to refactor to avoid doublicate code in encode time

---
Refers to: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs:170 in 2751cbf. [](commit_id = 2751cbf9cc8214a00ef8ed9f242b505ccc676e46, deletion_comment = False)

File: api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/EngineConfigurationApiModel.cs [hunk 7]
```diff
@@ -43,10 +43,10 @@ public class EngineConfigurationApiModel {
         public int? MaxMessageSize { get; set; }
 
         /// <summary>
-        /// Max IoT D2C outgress message buffer size
+        /// Max IoT D2C egress message queue
```
Comment (mregen) [inline api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/EngineConfigurationApiModel.cs:46]: >essage queue [](start = 32, length = 12)

is this a count?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs [hunk 8]
```diff
@@ -278,20 +291,31 @@ public class DataFlowProcessingEngine : IProcessingEngine, IDisposable {
...
             }
         }
 
+        /// <summary>
+        /// Gets the percentage of a number of values
+        /// in relation the total ValueChanges.
+        /// </summary>
+        /// <param name="valueCount"></param>
+        private string GetValuePercentage(double valueCount) =>
+            $"{valueCount / _messageTrigger.ValueChangesCount:P0}";
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs:308]: >_messageTrigger.ValueChangesCount [](start = 28, length = 33)

check for divide by zero?

File: modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Runtime/LegacyCliOptions.cs (hunk 2 above)
Comment (luiscantero) [reply modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Runtime/LegacyCliOptions.cs:297]: 2 GB basically to account for a connectivity shortage. I wanted to use half of the available or free memory, but that info is not easy to get in .NET Core in a way that works on all platforms (would require some testing). I will change it to 1 GB as a compromise :)

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/MonitoredItemMessageEncoder.cs (hunk 4 above)
Comment (luiscantero) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/MonitoredItemMessageEncoder.cs:91]: I moved it to a constant _encodedMessageSizeOverhead in DataFlowProcessingEngine.cs. The payload is [protocol-agnostic](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-construct#message-size), so we shouldn't need to reserve so much space for the header.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs (hunk 5 above)
Comment (luiscantero) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs:106]: Many people (incl. me) consider this a [best-practice](https://medium.com/@nikgraf/why-you-should-enforce-dangling-commas-for-multiline-statements-d034c98e36f8). There are multiple discussions about this on the Internet and C#, JavaScript and other languages have support for it and there is a reason for that.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs (hunk 6 above)
Comment (luiscantero) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs:92]: Same as above

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs (hunk 8 above)
Comment (luiscantero) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs:308]: Done

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Publisher/src/Services/PublisherJobService.cs (hunk 3 above)
Comment (luiscantero) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Publisher/src/Services/PublisherJobService.cs]: Done

File: modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Runtime/LegacyCliOptions.cs (hunk 2 above)
Comment (luiscantero) [reply modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Runtime/LegacyCliOptions.cs]: Done

File: api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/EngineConfigurationApiModel.cs (hunk 7 above)
Comment (luiscantero) [reply api/src/Microsoft.Azure.IIoT.Api/src/Publisher/Models/EngineConfigurationApiModel.cs:46]: It's a count: MaxEgressMessageQueue * 256 KB would be the max. size.

**Comment by luiscantero** [comment]
> > ```
> >     private IEnumerable<NetworkMessageModel> EncodeBatchAsUadp(
> > ```
> 
> [Minor] maybe the right time to refactor to avoid doublicate code in encode time
> 
> Refers to: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/MonitoredItemMessageEncoder.cs:167 in [2751cbf](https://github.com/Azure/Industrial-IoT/commit/2751cbf9cc8214a00ef8ed9f242b505ccc676e46). [](commit_id = [2751cbf](https://github.com/Azure/Industrial-IoT/commit/2751cbf9cc8214a00ef8ed9f242b505ccc676e46), deletion_comment = False)

I think it would be better as part of another PR. I see your point, however those are changes to core-functionality and would require some thorough testing.

**Comment by luiscantero** [comment]
Commit 
bcc7505 "MaxEgressMessageQueue default to 4096 bytes" should be messages, not bytes.

**Pull request merged**